File: surgeon/simulate_calls.py
import os
import json
import urllib.request
from datetime import datetime

API_KEY = os.environ.get("GEMINI_API_KEY")

if not API_KEY:
    raise ValueError("Please set GEMINI_API_KEY")


MODEL_URL = "https://generativelanguage.googleapis.com/v1beta/models/gemini-3-flash-preview:generateContent"


def call_gemini(prompt):

    data = {
        "contents": [
            {
                "parts": [{"text": prompt}]
            }
        ]
    }

    req = urllib.request.Request(
        MODEL_URL,
        data=json.dumps(data).encode("utf-8"),
        headers={
            "Content-Type": "application/json",
            "x-goog-api-key": API_KEY
        },
        method="POST"
    )

    with urllib.request.urlopen(req) as response:
        result = json.loads(response.read())

    return result["candidates"][0]["content"]["parts"][0]["text"]


def replace_placeholders(prompt, customer):
    replacements = {
        "{{customer_name}}": customer.get("name", "Customer"),
        "{{pending_amount}}": customer.get("pending_amount", "fifty thousand"),
        "{{closure_amount}}": customer.get("closure_amount", "thirty five thousand"),
        "{{settlement_amount}}": customer.get("settlement_amount", "twenty five thousand"),
        "{{dpd}}": customer.get("dpd", "180"),
        "{{pos}}": customer.get("closure_amount", "thirty five thousand"),
        "{{tos}}": customer.get("pending_amount", "fifty thousand"),
        "{{due_date}}": "next week",
        "{{today_date}}": "today",
        "{{today_day}}": "Monday",
        "{{loan_id}}": "12345",
        "{{lender_name}}": "DemoLender"
    }
    for placeholder, value in replacements.items():
        prompt = prompt.replace(placeholder, value)
    return prompt


def simulate_call(prompt, transcript):

    customer = transcript.get("customer", {})
    prompt = replace_placeholders(prompt, customer)

    borrower_messages = [
        t["text"]
        for t in transcript["transcript"]
        if t["speaker"] == "customer"
    ]

    conversation = []

    context = prompt + "\n\nYou are continuing a phone call."

    for msg in borrower_messages:

        llm_prompt = f"""
{context}

Customer: {msg}

Agent:
"""

        reply = call_gemini(llm_prompt)

        conversation.append({
            "customer": msg,
            "agent": reply.strip()
        })

        context += f"\nCustomer: {msg}\nAgent: {reply}"

    return conversation


def save_simulation(call_id, conversation):

    os.makedirs("restimulated_outputs", exist_ok=True)

    timestamp = datetime.now().strftime("%Y%m%d_%H%M%S")

    filename = f"restimulated_outputs/{call_id}_{timestamp}.json"

    with open(filename, "w", encoding="utf-8") as f:
        json.dump(conversation, f, indent=2, ensure_ascii=False)

    return filename

File: surgeon/simulate_calls.js
import fs from "fs";
import process from "process";

const API_KEY = process.env.GEMINI_API_KEY;

if (!API_KEY) {
  throw new Error("Please set GEMINI_API_KEY");
}

const MODEL_URL =
  "https://generativelanguage.googleapis.com/v1beta/models/gemini-3-flash-preview:generateContent";

// ---------------------------------------------------------------------------
// Gemini Call
// ---------------------------------------------------------------------------

async function callGemini(prompt) {
  const data = {
    contents: [
      {
        parts: [{ text: prompt }],
      },
    ],
  };

  const response = await fetch(MODEL_URL, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
      "x-goog-api-key": API_KEY,
    },
    body: JSON.stringify(data),
  });

  const result = await response.json();

  return result.candidates?.[0]?.content?.parts?.[0]?.text ?? "";
}

// ---------------------------------------------------------------------------
// Replace Placeholders
// ---------------------------------------------------------------------------

function replacePlaceholders(prompt, customer = {}) {
  const replacements = {
    "{{customer_name}}": customer.name || "Customer",
    "{{pending_amount}}": customer.pending_amount || "fifty thousand",
    "{{closure_amount}}": customer.closure_amount || "thirty five thousand",
    "{{settlement_amount}}":
      customer.settlement_amount || "twenty five thousand",
    "{{dpd}}": customer.dpd || "180",
    "{{pos}}": customer.closure_amount || "thirty five thousand",
    "{{tos}}": customer.pending_amount || "fifty thousand",
    "{{due_date}}": "next week",
    "{{today_date}}": "today",
    "{{today_day}}": "Monday",
    "{{loan_id}}": "12345",
    "{{lender_name}}": "DemoLender",
  };

  for (const [placeholder, value] of Object.entries(replacements)) {
    prompt = prompt.replaceAll(placeholder, value);
  }

  return prompt;
}

// ---------------------------------------------------------------------------
// Simulate Call
// ---------------------------------------------------------------------------

export async function simulateCall(prompt, transcript) {
  const customer = transcript.customer || {};

  prompt = replacePlaceholders(prompt, customer);

  const borrowerMessages = transcript.transcript
    .filter((t) => t.speaker === "customer")
    .map((t) => t.text);

  const conversation = [];

  let context = prompt + "\n\nYou are continuing a phone call.";

  for (const msg of borrowerMessages) {
    const llmPrompt = `
${context}

Customer: ${msg}

Agent:
`;

    const reply = await callGemini(llmPrompt);

    conversation.push({
      customer: msg,
      agent: reply.trim(),
    });

    context += `\nCustomer: ${msg}\nAgent: ${reply}`;
  }

  return conversation;
}

// ---------------------------------------------------------------------------
// Save Simulation
// ---------------------------------------------------------------------------

export function saveSimulation(callId, conversation) {
  fs.mkdirSync("restimulated_outputs", { recursive: true });

  const timestamp = new Date()
    .toISOString()
    .replace(/[-:]/g, "")
    .slice(0, 15);

  const filename = `restimulated_outputs/${callId}_${timestamp}.json`;

  fs.writeFileSync(filename, JSON.stringify(conversation, null, 2));

  return filename;
}